lib/model/model.go

```diff
@@ -1945,13 +1935,39 @@ func (m *Model) internalScanFolderSubdirs(ctx context.Context, folder string, su
 
 	runner.setState(FolderScanning)
 
-	fchan, err := scanner.Walk(ctx, scanner.Config{
+	subPrefixes := subDirs
+	if len(subPrefixes) == 0 {
+		subPrefixes = []string{""}
+	}
+	haveChan := make(chan *protocol.FileInfo)
+	dbCtx, dbCancel := context.WithCancel(ctx)
+	go func() {
+		for _, sub := range subPrefixes {
+			select {
+			case <-dbCtx.Done():
+				break
+			default:
+			}
+			fset.WithPrefixedHave(protocol.LocalDeviceID, sub, func(fi db.FileIntf) bool {
+				f := fi.(protocol.FileInfo)
+				select {
+				case haveChan <- &f:
+				case <-dbCtx.Done():
+					return false
+				}
+				return true
+			})
+		}
+		close(haveChan)
+	}()
+
+	rchan, err := scanner.Walk(ctx, scanner.Config{
 		Folder:                folderCfg.ID,
 		Subs:                  subDirs,
 		Matcher:               ignores,
 		BlockSize:             protocol.BlockSize,
 		TempLifetime:          time.Duration(m.cfg.Options().KeepTemporariesH) * time.Hour,
-		CurrentFiler:          cFiler{m, folder},
+		HaveChan:              haveChan,
```

Comment:
Perhaps it's better to pass the FileSet and just let the scanner deal with getting it whatever it needs?
This way both iteration loops (subdirs here, and subdirs in the scanner) are in the same place.

Reply:
To me It looked like it was deliberate that the scanner package does not use db at all, so I kept it that way. If there is nothing to it, I agree that it would be better to have both iterations in one place.

Reply:
As long as it takes a reasonable interface it and isn't a direct dependency on the database that's probably fine either way.